Attendance Registration and Approval

nicolasantolini opened the pull request:
Refactor attendance management in SessionsController

- Renamed `EditAttendance` to `EditStudentAttendance` and added `EditTeacherAttendance` in `SessionsController`.
- Updated attendance logic to differentiate between student and teacher attendance.
- Created a migration to update the `AttendanceRecord` table schema, renaming `Record` to `TeacherAttendance` and adding `StudentAttendance`, `StudentSubmitted`, and `TeacherSubmitted`.
- Updated `CoursesContextModelSnapshot` to reflect the new schema.
- Modified `AttendanceRecord` to include separate properties for student and teacher attendance.
- Changed attendance setting methods in `Session` to allow for granular control.
- Updated `AttendanceService` to call the correct attendance methods for students and teachers.

nicolasantolini committed:
Refactor attendance management in SessionsController

- Renamed `EditAttendance` to `EditStudentAttendance` and added `EditTeacherAttendance` in `SessionsController`.
- Updated attendance logic to differentiate between student and teacher attendance.
- Created a migration to update the `AttendanceRecord` table schema, renaming `Record` to `TeacherAttendance` and adding `StudentAttendance`, `StudentSubmitted`, and `TeacherSubmitted`.
- Updated `CoursesContextModelSnapshot` to reflect the new schema.
- Modified `AttendanceRecord` to include separate properties for student and teacher attendance.
- Changed attendance setting methods in `Session` to allow for granular control.
- Updated `AttendanceService` to call the correct attendance methods for students and teachers.

joachimdekker commented on AttendanceSystem.Api/Controllers/SessionsController.cs:
```diff
@@ -49,16 +49,24 @@ public async Task<IActionResult> GetSessionInfo([HttpTrigger(AuthorizationLevel.
-    
+
+    [Function($"{nameof(SessionsController)}-{nameof(EditTeacherAttendance)}")]
+    public async Task<IActionResult> EditTeacherAttendance([HttpTrigger(AuthorizationLevel.Anonymous, "put", Route = "courses/{courseId}/sessions/{sessionId:guid}/teacherattendance")] HttpRequest req, string courseId, Guid sessionId, [FromBody] UpdateAttendanceContract contract)
```
I think the resource `teacherattendance` is misleading, as this could mean the attendance of the teacher. Perhaps something like `approvals` or something is better. Another way (perhaps more cleaner but more technical advanced way) to do it, is to just use the same endpoint and look at the role of the bearer token to determine which path to take. (That would make your PR dependent on mine, so I think we can put this idea on the backlog for now, if you agree with the idea).

nicolasantolini replied:
I definitely agree. For the moment I will keep the 2 method naming them better

joachimdekker commented on AttendanceSystem.Api/Controllers/SessionsController.cs:
```diff
@@ -49,16 +49,24 @@ public async Task<IActionResult> GetSessionInfo([HttpTrigger(AuthorizationLevel.
+    public async Task<IActionResult> EditTeacherAttendance([HttpTrigger(AuthorizationLevel.Anonymous, "put", Route = "courses/{courseId}/sessions/{sessionId:guid}/teacherattendance")] HttpRequest req, string courseId, Guid sessionId, [FromBody] UpdateAttendanceContract contract)
+    {
+        _logger.LogInformation("C# HTTP trigger function processed a request.");
+        await _attendanceService.SetTeacherAttendance(sessionId, contract.UserId, contract.Kind);
```
Like I said in my other comment, I think `SetTeacherAttendance` is not a good name, we could do something like `SetTeacherApproval` or `SetActualApproval`. (If we take the role-based way described above, we still have to have separate domain functions, so this is still relevant).

joachimdekker commented on AttendanceSystem.Domain.Model/AttendanceRecord.cs:
```diff
@@ -4,5 +4,8 @@ public record AttendanceRecord
     public required SessionId SessionId { get; set; }
     public required StudentId StudentId { get; set; }
-    public required AttendanceKind Record { get; set; }
+    public required AttendanceKind StudentAttendance { get; set; }
+    public required DateTime? StudentSubmitted { get; set; }
+    public required AttendanceKind TeacherAttendance { get; set; }
+    public required DateTime? TeacherSubmitted { get; set; }
```
Like @alexanderserraris already mentioned, we could create an `AttendanceSubmittance` (please come up with a better name haha) Value Object (just a record or a struct) that encapsulates the AttendanceKind and the DateTime that it was submitted. Then, we can alter the DbContext and [make those properties 'owned' properties, which mean they will inline in the database.](https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#configuring-types-as-owned)

joachimdekker commented on AttendanceSystem.Domain.Services/AttendanceService.cs:
```diff
@@ -45,10 +45,17 @@ public async Task DeleteSession(Guid sessionId)
+
+    public async Task SetTeacherAttendance(Guid sessionId, string teacherId, AttendanceKind kind)
+    {
+        var session = await GetSession(sessionId);
```
Does this work? I think it does not save correctly. In the GetSession, we don't have any include that includes the Register. We may want to create a private method that gets the session but includes the register for this method, just for optimalisation reasons.

joachimdekker commented on AttendanceSystem.Domain.Model/Session.cs:
```diff
@@ -35,12 +35,32 @@ public Session(Course course, List<StudentId> students, DateTime startTime, Date
+
+    public void SetTeacherAttendance(StudentId studentId, AttendanceKind attendance, DateTime submitted)
+    {
+        var record = Register.FirstOrDefault(r => r.StudentId == studentId);
+        if (record is null)
+        {
+            throw new ArgumentException("Student is not registered for this session.");
+        }
```
I like the `??` better, but that is a matter of preference of course.
```suggestion
        var record = Register.FirstOrDefault(r => r.StudentId == studentId) ?? throw new ArgumentException("Student is not registered for this session.");
```

Same goes for the lines above.

joachimdekker commented on AttendanceSystem.Domain.Model/Session.cs:
```diff
@@ -35,12 +35,32 @@ public Session(Course course, List<StudentId> students, DateTime startTime, Date
-        Register.Single(r => r.StudentId == studentId).Record = kind;
+    public void SetStudentAttendance(StudentId studentId, AttendanceKind attendance, DateTime submitted)
+    {
+        var record = Register.FirstOrDefault(r => r.StudentId == studentId);
+        if (record is null)
+        {
+            throw new ArgumentException("Student is not registered for this session.");
+        }
```
```suggestion
        var record = Register.FirstOrDefault(r => r.StudentId == studentId) ?? throw new ArgumentException("Student is not registered for this session.");
```

joachimdekker reviewed (changes requested):
Nice PR, I left a few comments that also reflect what we discussed on WhatsApp with @alexanderserraris

nicolasantolini committed:
Update AttendanceSystem.Domain.Model/Session.cs

Co-authored-by: Joachim Dekker <jdekker3008@gmail.com>

nicolasantolini committed:
Update AttendanceSystem.Domain.Model/Session.cs

Co-authored-by: Joachim Dekker <jdekker3008@gmail.com>

nicolasantolini committed:
Refactor attendance management and update migrations

- Changed `EditTeacherAttendance` to use `SetTeacherApproval`.
- Added `AttendanceSubmission` model and updated migrations.
- Updated `AttendanceRecord` to reference `AttendanceSubmission`.
- Modified `Session` to use `AttendanceSubmission` for attendance records.
- Introduced `GetSessionWithRegister` method in `AttendanceService`.
- Updated project file to include `Microsoft.EntityFrameworkCore.Abstractions`.

nicolasantolini committed:
Rename attendance method in SessionsController

Removed the `EditStudentAttendance` method and added
`ConfirmStudentAttendance` in the `SessionsController` class.
Both methods handle PUT requests for updating student
attendance, but the new method may have altered functionality.

joachimdekker commented on AttendanceSystem.Domain.Model/AttendanceSubmission.cs:
```diff
@@ -0,0 +1,20 @@
+        public required AttendanceKind StudentAttendance { get; set; }
+        public required DateTime? StudentSubmitted { get; set; }
+        public required AttendanceKind TeacherAttendance { get; set; }
+        public required DateTime? TeacherSubmitted { get; set; }
```
Ah, you did not understand me well. I want to have a single AttendanceSubmission which encompasses the AttendanceKind and the DateTime as one single object. Then, in the AttendanceRecord, we can have two AttendanceSubmission objects.

joachimdekker committed:
refactor: attendance submission model and logic

Introduced separate `StudentSubmission` and `TeacherSubmission` in `AttendanceRecord` to replace shared `AttendanceSubmission`. Enhanced logic to record submission timestamps automatically and streamlined methods to remove redundant parameters. Updated project settings to use C# preview features.

nicolasantolini committed:
Merge branch 'main' into AttendanceApproval

joachimdekker committed:
Merge remote-tracking branch 'origin/main' into AttendanceApproval

# Conflicts:
#	AttendanceSystem.Api/Controllers/SessionsController.cs

joachimdekker committed:
fix: ensure proper materialization of student list

Fixes an issue where the student list was not being fully materialized, potentially leading to deferred execution issues. Also includes minor formatting adjustments for consistency.

joachimdekker committed:
refactor: add SwaggerIgnore to FunctionContext parameters

Added the [SwaggerIgnore] attribute to FunctionContext parameters in various controllers. This prevents these parameters from being included in the Swagger documentation, improving its clarity and relevance to API consumers.

alexanderserraris reviewed (approved):
LGTM!
